File: ga/gpt3_5_result_2024_6/result/1968G2_gpt/10_1968G2.c
use std::io::{self, BufRead};

const N: usize = 200000;
const N_: usize = 1 << 18; // N_ = pow2(ceil(log2(N)))

fn max(a: usize, b: usize) -> usize {
    if a > b {
        a
    } else {
        b
    }
}

fn build(aa: &Vec<usize>, n: usize) -> Vec<usize> {
    let mut st = vec![0; N_ * 2];
    let mut n_ = 1;
    
    while n_ < n {
        n_ <<= 1;
    }

    for i in 0..n {
        st[n_ + i] = aa[i];
    }

    for i in (1..n_).rev() {
        st[i] = max(st[i << 1], st[i << 1 | 1]);
    }

    st
}

fn query(l: usize, a: usize, st: &Vec<usize>, n_: usize) -> usize {
    let mut l = l + n_;
    let mut r = n_ - 1;

    while l <= r {
        if l & 1 == 1 {
            if st[l] >= a {
                let mut temp = l;
                while temp < n_ {
                    temp = if st[temp << 1] >= a { temp << 1 } else { temp << 1 | 1 };
                }
                return temp - n_;
            }
            l += 1;
        }
        l >>= 1;
        r >>= 1;
    }

    n_
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines().map(|line| line.unwrap());
    
    let t: usize = lines.next().unwrap().parse().unwrap();

    for _ in 0..t {
        let input: Vec<usize> = lines
            .next()
            .unwrap()
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        
        let n = input[0];
        let kl = input[1];
        let kr = input[2];
        let cc = input[3].to_string();

        let mut zz = vec![0; n];
        let mut l = 0;
        let mut r = 0;

        for i in 1..n {
            if i + zz[i - l] < r {
                zz[i] = zz[i - l];
            } else {
                l = i;
                r = max(r, i);
                while r < n && cc.chars().nth(r) == cc.chars().nth(r - l) {
                    r += 1;
                }
                zz[i] = r - l;
            }
        }

        let st = build(&zz, n);
        let mut ans = vec![0; n + 1];

        for p in 1..=n {
            let mut k = 0;
            let mut i = 0;
            while i < n {
                i = query(i + p, p, &st, n_);
                if i != n_ {
                    k += 1;
                }
            }
            ans[k] = max(ans[k], p);
        }

        for k in (1..n).rev() {
            ans[k] = max(ans[k], ans[k + 1]);
        }

        for k in kl..=kr {
            print!("{} ", ans[k]);
        }
        println!();
    }
}

File: ga/4omini_result_2024_7/result/1968G2_mini/10_1968G2.rs
use std::io::{self, BufRead};

fn max(a: usize, b: usize) -> usize {
    if a > b { a } else { b }
}

struct SegmentTree {
    st: Vec<usize>,
    n: usize,
}

impl SegmentTree {
    fn new(size: usize) -> Self {
        let n = size.next_power_of_two();
        SegmentTree {
            st: vec![0; n * 2],
            n,
        }
    }

    fn build(&mut self, aa: &[usize]) {
        self.st[self.n..self.n + aa.len()].copy_from_slice(aa);
        for i in (1..self.n).rev() {
            self.st[i] = max(self.st[i * 2], self.st[i * 2 + 1]);
        }
    }

    fn query(&self, mut l: usize, mut a: usize) -> usize {
        let mut r = self.n - 1;
        let (mut l, mut r) = (l + self.n, r + self.n);

        while l <= r {
            if l & 1 == 1 {
                if self.st[l] >= a {
                    let mut result = l;
                    while result < self.n {
                        result = if self.st[result * 2] >= a {
                            result * 2
                        } else {
                            result * 2 + 1
                        };
                    }
                    return result - self.n;
                }
                l += 1;
            }
            l >>= 1;
            r >>= 1;
        }
        self.n
    }
}

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();

    let mut inputs = reader.lines().map(|x| x.unwrap());
    let t: usize = inputs.next().unwrap().parse().unwrap();

    for _ in 0..t {
        let header: Vec<&str> = inputs.next().unwrap().split_whitespace().collect();
        let n: usize = header[0].parse().unwrap();
        let kl: usize = header[1].parse().unwrap();
        let kr: usize = header[2].parse().unwrap();
        let cc: String = inputs.next().unwrap();

        let mut zz = vec![0; n];
        let (mut l, mut r) = (0, 0);
        
        for i in 1..n {
            if i + zz[i - l] < r {
                zz[i] = zz[i - l];
            } else {
                l = i;
                r = max(r, i);
                while r < n && cc.chars().nth(r) == cc.chars().nth(r - l) {
                    r += 1;
                }
                zz[i] = r - l;
            }
        }

        let mut segment_tree = SegmentTree::new(n);
        segment_tree.build(&zz);

        let mut ans = vec![0; n + 1];
        for p in 1..=n {
            let mut k = 0;
            let mut i = 0;

            while i < n {
                i = segment_tree.query(i + p, p);
                k += 1;
            }
            ans[k] = max(ans[k], p);
        }

        for k in (1..=n - 1).rev() {
            ans[k] = max(ans[k], ans[k + 1]);
        }

        let result: Vec<String> = (kl..=kr).map(|k| ans[k].to_string()).collect();
        println!("{}", result.join(" "));
    }
}